src/frontend/src/expr/window_function.rs

```diff
@@ -0,0 +1,153 @@
+#[derive(Clone, Eq, PartialEq, Hash)]
+pub struct WindowFunction {
+    pub args: Vec<ExprImpl>,
+    pub return_type: DataType,
+    pub function_type: WindowFunctionType,
+    pub partition_by: Vec<ExprImpl>,
+    pub order_by: Vec<(ExprImpl, Direction)>,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum WindowFunctionType {
+    RowNumber,
+    Rank,
+    DenseRank,
+}
+
+impl WindowFunctionType {
+    pub fn name(&self) -> &str {
+        match self {
+            WindowFunctionType::RowNumber => "row_number",
+            WindowFunctionType::Rank => "rank",
+            WindowFunctionType::DenseRank => "dense_rank",
+        }
+    }
+}
+
+impl FromStr for WindowFunctionType {
+    type Err = ErrorCode;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        if s.eq_ignore_ascii_case("row_number") {
+            Ok(WindowFunctionType::RowNumber)
+        } else if s.eq_ignore_ascii_case("rank") {
+            Ok(WindowFunctionType::Rank)
+        } else if s.eq_ignore_ascii_case("dense_rank") {
+            Ok(WindowFunctionType::DenseRank)
+        } else {
+            Err(ErrorCode::NotImplemented(
+                format!("unknown table function kind: {s}"),
+                None.into(),
+            ))
+        }
+    }
+}
+
+impl WindowFunction {
+    /// Create a `WindowFunction` expr with the return type inferred from `func_type` and types of
+    /// `inputs`.
+    pub fn new(
+        function_type: WindowFunctionType,
+        partition_by: Vec<ExprImpl>,
+        order_by: Vec<(ExprImpl, Direction)>,
+        args: Vec<ExprImpl>,
+    ) -> Result<Self> {
+        if !args.is_empty() {
+            return Err(ErrorCode::BindError(format!(
+                "the length of args of {} function should be 0",
+                function_type.name()
+            ))
+            .into());
+        }
+
+        Ok(Self {
+            args,
+            return_type: DataType::Int64,
+            function_type,
+            partition_by,
+            order_by,
+        })
+    }
+}
+
+impl std::fmt::Debug for WindowFunction {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if f.alternate() {
+            f.debug_struct("WindowFunction")
+                .field("function_type", &self.function_type)
+                .field("return_type", &self.return_type)
+                .field("args", &self.args)
+                .field("partition_by", &self.partition_by)
+                .field("order_by", &self.order_by)
+                .finish()
+        } else {
+            let func_name = format!("{:?}", self.function_type);
+            let mut builder = f.debug_tuple(&func_name);
+            self.args.iter().for_each(|child| {
+                builder.field(child);
+            });
+            builder.finish()?;
+
+            if !self.partition_by.is_empty() || !self.order_by.is_empty() {
```

Comment:
This check seems unnecessary. `over()` is required even when both are empty, so that it can be distinguished from agg calls. In PostgreSQL:
```
test=# select rank() from t;
ERROR:  window function rank requires an OVER clause
LINE 1: select rank() from t;
               ^

test=# select rank() over () from t;
```